modules/bash-commons/src/aws.sh

```diff
@@ -3,26 +3,200 @@
+# The endpoint returns a token that must be supplied on subsequent requests. This implementation fetches a new token
+# for each transaction. See:
+# https://aws.amazon.com/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/
+# for more information
+function ec2_metadata_http_get {
+  assert_not_empty "path" "$1"
+  local -r path="$1"
+  # We allow callers to configure the ttl - if not provided it will default to 6 hours
+  local ttl=""
+  ttl=$(configure_imdsv2_ttl "$2")
+  token=$(ec2_metadata_http_put "$ttl")
+  curl "$metadata_endpoint/meta-data/$path" -H "X-aws-ec2-metadata-token: $token" \
+    --silent --location --fail --show-error
+}
+
+# This function uses Instance Metadata service version 2. It requests the supplied path from the
+# dynamic endpoint.
+function ec2_metadata_dynamic_http_get {
+  assert_not_empty "path" "$1"
+  local -r path="$1"
+  # We allow callers to configure the ttl - if not provided it will default to 6 hours
+  local ttl=""
+  ttl=$(configure_imdsv2_ttl "$2")
+  token=$(ec2_metadata_http_put $three_hours_in_s)
+  curl "$metadata_dynamic_endpoint/$path" -H "X-aws-ec2-metadata-token: $token" \
+    --silent --location --fail --show-error
+}
+
+# This function uses Instance Metadata Service version 2. It retrieves a token from the IMDSv2
+# endpoint, which can be presented during subequent requests to the IMDSv2 endpoints.
+
+# This function accepts a conifgurable TTL for the IMDSv2 token it requests. If a TTL is not supplied,
+# this function will default to the maximum IMDSv2 session duration which is 6 hours.
+function ec2_metadata_http_put {
+  # We allow callers to configure the ttl - if not provided it will default to 6 hours
+  local ttl=""
+  ttl=$(configure_imdsv2_ttl "$2")
+  token=$(curl --silent --location --fail --show-error -X PUT "$metadata_endpoint/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: $ttl")
+  echo "$token"
+}
+
+# This function uses Instance Metadata Service version 2. It retrieves the field of the supplied name
+# from the Instance Metadata Service's identity document for the given EC2 instance and returns
+# its value.
+#
+# This function uses Instance Metadata Service version 2. It accepts a configurable TTL
+# for the IMDSv2 token it requests, and it returns the token to the caller. If a TTL
+# is not supplied, this function will default to the maximum IMDSv2 session duration which is 6 hours.
+function ec2_instance_identity_field_get {
+  local -r field="$1"
+  local ttl=""
+  ttl=$(configure_imdsv2_ttl "$2")
+  token=$(ec2_metadata_http_put "$ttl")
+  curl "$instance_identity_endpoint" -H "X-aws-ec2-metadata-token: $token" \
+    --silent --location --fail --show-error | jq -r ".${field}"
+}
+
+# This is a convenience function for setting a TTL, and falling back to sensible defaults
+# when the value is either not supplied or out of bounds.
```

Comment:
Should the behavior be to throw an error when out of bounds instead of falling back to the default? If we just ignore the invalid value and move on, a user might mistakenly believe that an invalid value is valid.

Reply:
I added an error message in 89c1c88 - so the behavior is now: if a caller supplies a TTL that is greater than the 6 hour limit for the token, a warning error message will be printed explaining that the value is excessive and that the fallback of 6 hours is being used. 

I think this is a decent middle road - if we add an explicit exit this becomes much harder to test with the current harness.